src/orange/surf/SoftSurfaceEqual.cc

```diff
@@ -0,0 +1,173 @@
+    return soft_eq_(a.position(), b.position());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(PlaneAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two centered axis-aligned cylinders for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(CylCentered<T> const& a,
+                                  CylCentered<T> const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(CylCentered);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two centered spheres for near equality.
+ */
+bool SoftSurfaceEqual::operator()(SphereCentered const& a,
+                                  SphereCentered const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two aligned cylinders for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(CylAligned<T> const& a,
+                                  CylAligned<T> const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq())
+           && soft_eq_distance_(a.calc_origin(), b.calc_origin());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(CylAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two planes for near equality.
+ */
+bool SoftSurfaceEqual::operator()(Plane const& a, Plane const& b) const
+{
+    // Guard against dot product being slightly greater than 1 due to fp
+    // arithmetic
+    real_type const ndot = dot_product(a.normal(), b.normal());
+    return (ndot >= 1 || std::sqrt(1 - ndot) < soft_eq_.rel() * 3)
+           && soft_eq_(a.displacement(), b.displacement());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two spheres for near equality.
+ */
+bool SoftSurfaceEqual::operator()(Sphere const& a, Sphere const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq())
+           && soft_eq_distance_(a.origin(), b.origin());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two cones for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(ConeAligned<T> const& a,
+                                  ConeAligned<T> const& b) const
+{
+    return soft_eq_sq_(a.tangent_sq(), b.tangent_sq())
+           && soft_eq_distance_(a.origin(), b.origin());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(ConeAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two simple quadrics for near equality.
+ */
+bool SoftSurfaceEqual::operator()(SimpleQuadric const& a,
+                                  SimpleQuadric const& b) const
+{
+    return soft_eq_distance_(make_array(a.second()), make_array(b.second()))
```

Comment:
Why are the squared terms being compared with a linear comparator?

Reply:
They're being compared as vector terms: the distance comparison here does a dot product and a sqrt of that. It seems to pass the unit tests nicely.